### ISSUE TITLE
Clarify revert and regression policy in the case where new PRs are merged on top of the "problematic" PR

### ISSUE THREAD

**Issue by seanlip** [oppia/oppia-web-developer-docs#52]
Per discussion with @heyimShivam, it would be good if this page:

https://github.com/oppia/oppia/wiki/Revert-and-Regression-Policy 

had additional clarification on the procedure for fixing regressions in the explicit case where new PRs are merged on top of the PR that caused the regression.

------

(My take is that we should still revert the bad PR, and also try to enforce small PRs that are easily revertable in general. But open to other thoughts here, before we write this down as the rule. I'm also open to anything that keeps develop clean -- maybe we can say that the developer who introduced the issue should be contacted and must fix it within 24 hours, otherwise the revert will happen, and the revert might happen earlier than that at the discretion of the tech lead -- but again, open to suggestions here.)

**Comment by U8NWXD** [comment]
@gp201 and I discussed this. There are 2 ways to fix a bad PR: reverting the PR and merging another PR to fix the problem ("fixing forward"). Depending on the situation, both can be fast or slow. Reverting can be quick if there are no merge conflicts, but with large PRs, reverts can be hard. Fixing forward can be easy if it's a small, easily-identifiable bug, but hard if the solution is unclear. We should generally take whichever approach will fix develop first, possibly by pursuing both in parallel.

However, we both also have concerns that this focus on fixing develop can be overly harsh when developers have deadlines, e.g. for GSoC. Here's a loaded example to illustrate our concern: imagine a situation where a student merges a PR that's integral to their project close to the GSoC deadline. In that PR, there's a small bug that causes a minor regression on develop that causes minor issues for developers. The bug can be easily fixed forward and represents a small piece of the overall PR, but reverting would be clean and slightly faster than fixing forward. In this situation, reverting seems unnecessarily harsh since it only marginally benefits developers but severely harms the student.

**Comment by seanlip** [comment]
In the second case, would it be correct to say that one of the following
things should happen?

(a) The bug is deemed breaking and must be fixed forward, within X hours,
where X is a threshold that the dev workflow leads should determine in this
policy (and also should determine whether this refers to the PR being
opened or the PR being merged)

(b) The dev workflow leads decide to accept this as a permanent regression
on develop (because it's not serious or affecting folks much) and file an
issue for it to be fixed later. I think this is more possible with issues
that affect the dev team than regressions for users, though.

Thoughts?

**Comment by seanlip** [comment]
In other words, there is still a hard rule that "develop should not be
broken". What this means is fairly clear for end users, but could
potentially be a bit flexible for developers. For that latter category, it
seems reasonable to me for the dev workflow leads to be responsible for
deciding what "not broken" means.

**Comment by U8NWXD** [comment]
I think that makes sense, though I think X should be a function of severity. How about this policy:

-----------

When a PR is merged that causes a regression in develop, there are two ways to fix the regression:

1. Revert the PR. This can be quick if there are no merge conflicts, but with large PRs, reverts can be difficult.
2. Fix forward by merging another PR without reverting the problematic PR. This can be easy if the bug is small and easily-identifiable, but if the solution is unclear, fixing forward can be hard.

By default, we should take whichever approach will fix develop first, possibly by pursuing both in parallel.

However, the dev workflow team leads may decide to make the following exceptions to this policy:

1. If the regression in develop has only a minor detrimental effect on developers, we may decide to accept it as a permanent regression. Moving forward, the new "broken" state would be considered the intended state, and no fix would be required. An issue should be opened describing how the problem can be fixed, and this issue will be handled by the dev workflow team like any other feature request.

3. If the regression in develop (a) affects only developers, (b) the regression does not severely hamper developers' work, and (c) a revert would substantially harm the PR author (e.g. an impending GSoC deadline), we may allow the PR author to fix the regression forward even if reverting the PR would fix develop faster. In making this exception, we should set a clear deadline for the fix forward based on the severity of the regression:
   1. If the regression substantially impedes development, the fix forward PR should be merged within 2 days.
   2. If the regression only slightly impedes development, the fix forward PR should be merged within 7 days.

**Comment by seanlip** [comment]
Hmm, this looks like a good backbone but I have the following questions:

- What is the difference between the conditions in 1 and 2, and in which
circumstance does either apply? Is "minor detrimental effect" the same as
"does not severely hamper developers' work"?

- In 2, how can a regression "substantially impede development" while not
"severely hampering developers' work"?

I wonder if this is more of a two-stage process: dev workflow leads decide
which bucket to put the regression in (don't care / can live without for a
week / problematic), and then it falls under the policy for that bucket.

Also, two suggestions:
- I would change "GSoC" to "external", it seems odd for GSoC to show up in
a general policy like this
- You might also want to clarify that user-facing regressions are subject
to the baseline policy of "fix asap".

**Comment by U8NWXD** [comment]
@seanlip thanks for pointing out those ambiguities. Is this more clear?

-----------

When a PR is merged that causes a regression in develop, there are two ways to fix the regression:

1. Revert the PR. This can be quick if there are no merge conflicts, but with large PRs, reverts can be difficult.
2. Fix forward by merging another PR without reverting the problematic PR. This can be easy if the bug is small and easily-identifiable, but if the solution is unclear, fixing forward can be hard.

Note that in either case, the PR author is responsible for fixing the bug in their PR, either by re-creating their PR without the bug (if we revert) or by opening a new PR to fix the bug (if we fix forward).

By default, we should take whichever approach will fix develop first, possibly by pursuing both in parallel.

However, the dev workflow team leads may decide to make the following exceptions to this policy:

1. If the regression in develop has only a minor detrimental effect on developers, we may decide to accept it as a permanent regression. Moving forward, the new "broken" state would be considered the intended state, and no fix would be required. An issue should be opened describing how the problem can be fixed, and this issue will be handled by the dev workflow team like any other feature request.

2. If (a) the regression in develop affects only developers, (b) the regression does not have a severe detrimental effect on developers, and (c) a revert would substantially harm the PR author (e.g. an impending external deadline), we may allow the PR author to fix the regression forward even if reverting the PR would fix develop faster. In making this exception, we should set a clear deadline for the fix forward PR to be merged (2-7 days) based on the severity of the regression.

Here is a table showing when each exception applies:

| Detrimental Effect on Developers | Applicable Exceptions |
|---------|--------|
| Severe   | None   |
| Moderate | 2 (if points 2a and 2c are met)     |
| Minor    |  1, 2 (if points 2a and 2c are met)     |

Note that if a regression affects anyone besides developers, neither exception applies and we should take the default approach of fixing develop as soon as possible.

**Comment by seanlip** [comment]
Hi @U8NWXD, thanks! I think this is better but I still have some questions:

- "By default, we should take whichever approach will fix develop first, possibly by pursuing both in parallel." Who specifically (or which subset of people) has the responsibility for making the decision on whether to revert, fix forward, or pursue both? This is important to call out since that person is basically going to be the contact point for things like this when they arise.
- Similarly, who decides on whether exception (1) applies?
- The table at the bottom is a bit odd in calling out just points 2a and 2c. Is 2b not relevant? It seems odd to include it in the "if" clause but not in the table.

More generally, I think what could be helpful is to have sections here for:
- "If you are a developer and you see an error in develop that's blocking you from submitting your PR, do this"
- "If you are a developer and you've been notified that your PR has caused a breakage in develop, do this"
- "If you are a dev workflow team lead (or representative) and have received a complaint about a breakage in develop, do this"

and then someone who falls into the second category (say) can just follow the steps there, which should include clear points of contact and escalation. This would be similar to how we describe our [triage process](https://github.com/oppia/oppia/wiki/Triaging-process), which can be followed more directly as a "recipe" (and I suspect that writing it that way will help squeeze out all the remaining ambiguity and potentially simplify the process).

**Comment by U8NWXD** [comment]
> - "By default, we should take whichever approach will fix develop first, possibly by pursuing both in parallel." Who specifically (or which subset of people) has the responsibility for making the decision on whether to revert, fix forward, or pursue both? This is important to call out since that person is basically going to be the contact point for things like this when they arise.

This is determined by the dev workflow leads.

> - Similarly, who decides on whether exception (1) applies?

The policy states that the dev workflow leads decide whether to make exceptions, so they also decide whether the exceptions apply. Note that even if an exception applies, the dev workflow leads might decide to follow the standard "fix develop asap" policy.

> - The table at the bottom is a bit odd in calling out just points 2a and 2c. Is 2b not relevant? It seems odd to include it in the "if" clause but not in the table.

Point 2b is "the regression does not have a severe detrimental effect on developers", which is reflected in the first column.

> More generally, I think what could be helpful is to have sections here for:
> - "If you are a developer and you see an error in develop that's blocking you from submitting your PR, do this"
> - "If you are a developer and you've been notified that your PR has caused a breakage in develop, do this"
> - "If you are a dev workflow team lead (or representative) and have received a complaint about a breakage in develop, do this"



> and then someone who falls into the second category (say) can just follow the steps there, which should include clear points of contact and escalation. This would be similar to how we describe our [triage process](https://github.com/oppia/oppia/wiki/Triaging-process), which can be followed more directly as a "recipe" (and I suspect that writing it that way will help squeeze out all the remaining ambiguity and potentially simplify the process).

done

Here's an updated policy:

-----------

## Policy

When a PR is merged that causes a regression in develop, there are two ways to fix the regression:

1. Revert the PR. This can be quick if there are no merge conflicts, but with large PRs, reverts can be difficult.
2. Fix forward by merging another PR without reverting the problematic PR. This can be easy if the bug is small and easily-identifiable, but if the solution is unclear, fixing forward can be hard.

Note that in either case, the PR author is responsible for fixing the bug in their PR, either by re-creating their PR without the bug (if we revert) or by opening a new PR to fix the bug (if we fix forward).

By default, we should take whichever approach will fix develop first, possibly by pursuing both in parallel. The dev workflow leads are responsible for deciding which approach(es) to follow.

However, the dev workflow team leads may decide to make the following exceptions to this policy:

1. If the regression in develop has only a minor detrimental effect on developers, we may decide to accept it as a permanent regression. Moving forward, the new "broken" state would be considered the intended state, and no fix would be required. An issue should be opened describing how the problem can be fixed, and this issue will be handled by the dev workflow team like any other feature request.

2. If (a) the regression in develop affects only developers, (b) the regression does not have a severe detrimental effect on developers, and (c) a revert would substantially harm the PR author (e.g. an impending external deadline), we may allow the PR author to fix the regression forward even if reverting the PR would fix develop faster. In making this exception, we should set a clear deadline for the fix forward PR to be merged (2-7 days) based on the severity of the regression.

Here is a table showing when each exception applies:

| Detrimental Effect on Developers | Applicable Exceptions |
|---------|--------|
| Severe   | None   |
| Moderate | 2 (if points 2a and 2c are met)     |
| Minor    |  1, 2 (if points 2a and 2c are met)     |

Note that if a regression affects anyone besides developers, neither exception applies and we should take the default approach of fixing develop as soon as possible.

## Instructions

### Developers Who Notice a Regression

If you see a regression in develop, notify the dev workflow team by opening an issue (the "regression tracking issue") that describes the regression in detail. Please include:

* A description of the previous (correct) behavior.
* A description of the current (broken) behavior.
* What impact the regression has on affected stakeholders (e.g. learners, developers, exploration creators, etc.).
* A discussion of any investigations you have done and the cause of the regression, if known.

Please follow-up with an email to dev workflow team members so that we can get to work fixing the breakage quickly.

### Developers Who Cause a Regression

If you have been notified that your changes caused a regression in develop, you should first reach out to the dev workflow team via the tracking issue so they can advise you on which remediation approach should be followed. If no tracking issue (discussed in the previous section) exists, you should create it. In your comment, please estimate how quickly you can merge a PR to fix the regression (fix forward).

You should also begin debugging your changes to figure out what caused the regression and how to fix it. Regardless of whether we end up reverting the PR or fixing forward, you're responsible for fixing the bug, either by re-creating the PR after the revert or by fixing forward.

Once the dev workflow team decides whether to revert, fix forward, or both, you should start work on the associated actions (re-creating the PR with the bug fixed, opening a fix-forward PR to fix the bug, or both).

### Dev Workflow Team

Once you are informed of a regression in develop, you should make sure a tracking issue is created and alert the rest of the dev workflow team. Work with the team leads to decide, following the above policy, whether to revert the PR, fix forward, pursue both in parallel, or make one of the two allowed exceptions. Once a decision is made, communicate it in the tracking issue and execute it (i.e. by reverting the PR or making sure the author fixes forward).

**Comment by seanlip** [comment]
Thanks @U8NWXD, this looks great! Some thoughts for the bottom three sections:

- Developers Who Notice a Regression
    - Perhaps worth making an issue template for regressions with the fields in question.
    - Also, could you specify the email address for the "email to dev workflow team members"? Also is this team members or leads?
- Developers Who Cause a Regression
    - When you say "reach out to the dev workflow team" do you mean the team leads instead? This is for both the first and third paragraphs.
- Dev Workflow Team
    - Ah, this is interesting because it looks like the whole dev workflow team is getting involved. Just to confirm, this would include folks working on typing, tests, linter, etc. -- i.e. the aim is that this is "basic training" for anyone in the dev workflow team? If so, that sounds fine to me but I just wanted to double-check because I know we have some folks in the team working on CI stuff and some who aren't.

For the top section, I think part of my earlier confusion was because I interpreted "not severe" to be "minor". Might I suggest that we have an explicit step that classifies the regression as minor, moderate or severe? Currently I don't think it's really stated explicitly that there are three levels of categorization (except "in passing" in the summary table). Once the issue is categorized into the correct bucket, it becomes quite clear what the allowed options are. I would also suggest swapping 2a and 2b so that the severe/moderate/minor triage criterion comes first.

**Comment by U8NWXD** [comment]
I agree with creating an issue template. I'll do that once we settle on what the policy statement (below) should be.

Updated based on comments and offline discussion:

-----------

## Policy

When a PR is merged that causes a regression in develop, there are two ways to fix the regression:

1. Revert the PR. This can be quick if there are no merge conflicts, but with large PRs, reverts can be difficult.
2. Fix forward by merging another PR without reverting the problematic PR. This can be easy if the bug is small and easily-identifiable, but if the solution is unclear, fixing forward can be hard.

Note that in either case, the PR author is responsible for fixing the bug in their PR, either by re-creating their PR without the bug (if we revert) or by opening a new PR to fix the bug (if we fix forward).

By default, we should take whichever approach will fix develop first, possibly by pursuing both in parallel. The dev workflow leads are responsible for deciding which approach(es) to follow.

However, based in part on the severity (minor, moderate, or severe) of the regression, the dev workflow team leads may decide to make the following exceptions to this policy:

1. If the regression in develop has only a minor detrimental effect on developers, we may decide to accept it as a permanent regression. Moving forward, the new "broken" state would be considered the intended state, and no fix would be required. An issue should be opened describing how the problem can be fixed, and this issue will be handled by the dev workflow team like any other feature request.

2. If (a) the regression does not have a severe detrimental effect on developers, (b) the regression in develop affects only developers, and (c) a revert would substantially harm the PR author (e.g. an impending external deadline), we may allow the PR author to fix the regression forward even if reverting the PR would fix develop faster. In making this exception, we should set a clear deadline for the fix forward PR to be merged (2-7 days) based on the severity of the regression.

To decide whether an exception is available, first decide whether anyone besides developers are affected. If non-developers are affected, no exceptions are available. Next, decide whether the regression has a minor, moderate, or severe detrimental effect on developers. Then this table shows when each exception applies:

| Detrimental Effect on Developers | Applicable Exceptions |
|---------|--------|
| Severe   | None   |
| Moderate | 2 (if points 2b and 2c are met)     |
| Minor    |  1, 2 (if points 2b and 2c are met)     |

## Instructions

### Developers Who Notice a Regression

If you see a regression in develop, notify the dev workflow leads by opening an issue (the "regression tracking issue") that describes the regression in detail. Please include:

* A description of the previous (correct) behavior.
* A description of the current (broken) behavior.
* What impact the regression has on affected stakeholders (e.g. learners, developers, exploration creators, etc.).
* A discussion of any investigations you have done and the cause of the regression, if known.

Please at-mention @oppia/dev-workflow-reviewers so that we can get to work fixing the breakage quickly.

### Developers Who Cause a Regression

If you have been notified that your changes caused a regression in develop, you should first reach out to the dev workflow leads via the tracking issue so they can advise you on which remediation approach should be followed. If no tracking issue (discussed in the previous section) exists, you should create it. In your comment, please estimate how quickly you can merge a PR to fix the regression (fix forward).

You should also begin debugging your changes to figure out what caused the regression and how to fix it. Regardless of whether we end up reverting the PR or fixing forward, you're responsible for fixing the bug, either by re-creating the PR after the revert or by fixing forward.

Once the dev workflow leads decide whether to revert, fix forward, or both, you should start work on the associated actions (re-creating the PR with the bug fixed, opening a fix-forward PR to fix the bug, or both).

### Dev Workflow Leads

Once you are informed of a regression in develop, you should make sure a tracking issue is created and alert the rest of the dev workflow leads. Decide, following the above policy, whether to revert the PR, fix forward, pursue both in parallel, or make one of the two allowed exceptions. Once a decision is made, communicate it in the tracking issue and execute it (i.e. by reverting the PR or making sure the author fixes forward).

**Comment by seanlip** [comment]
This looks great, thanks @U8NWXD! 

I only have one minor quibble: I suggest you swap 1 and 2. That way, the last row in the table reads as "1 (if points 1b and 1c are met), 2" which avoids the slight ambiguity of the current presentation where points 2b and 2c could conceivably apply to 1 as well. (This makes more sense prioritization-wise to me too because accepting a "permanent regression" feels more like it should be a backup option.)

But otherwise, lgtm :) (Also /cc @heyimShivam in case he has any comments, since I think he brought this up initially.)

**Comment by heyimShivam** [comment]
LGTM too. Thank @seanlip @U8NWXD @gp201 to make everything crystal clear.

**Comment by heyimShivam** [comment]
just a small question: what will happen when the user merges PR 1 back 3 months ago (before mid-term evaluation GSoC), and he also merges 3-4 pr on top of that, there was a regression caused in develop by the pr 1 (mid evaluation PR) and no one identified it till 3months, and now this is identified on last week of gsoc final evaluation.


in such a case I think he should get x no. of days to fix the pr as his whole work is dependent of pr 1, without reverting the pr.

**Comment by seanlip** [comment]
Hi @heyimShivam, I think this is probably more of a question for @U8NWXD (as dev workflow lead and GSoC admin). It's a good test case for the triage process, so I'll defer to him to answer it!

**Comment by U8NWXD** [comment]
@heyimShivam I think it would depend on how hard reverting the PR is versus fixing forward. With 3 months having gone by since PR 1, reverting is probably going to be hard if not impossible, so I expect fixing forward would be faster.

However, if for some reason fixing forward were harder than reverting, then we would check whether either exception applies. To get at the point I think you're making, let's assume that the regression is severe and blocks all development of oppia-web, so neither exception applies. (I could imagine this going unnoticed for 3 months if a PR were written so it only worked in 2022, so we didn't see the breakage until January 2023.) In this case, even though the contributor's whole project depends on PR 1, that PR is severely hampering development, so I think we have to revert. We can't save one person's GSoC project at the expense of development across Oppia. What do you think?

**Comment by heyimShivam** [comment]
> @heyimShivam I think it would depend on how hard reverting the PR is versus fixing forward. With 3 months having gone by since PR 1, reverting is probably going to be hard if not impossible, so I expect fixing forward would be faster.
> 
> However, if for some reason fixing forward were harder than reverting, then we would check whether either exception applies. To get at the point I think you're making, let's assume that the regression is severe and blocks all development of oppia-web, so neither exception applies. (I could imagine this going unnoticed for 3 months if a PR were written so it only worked in 2022, so we didn't see the breakage until January 2023.) In this case, even though the contributor's whole project depends on PR 1, that PR is severely hampering development, so I think we have to revert. We can't save one person's GSoC project at the expense of development across Oppia. What do you think?

Yes, I totally agree with you, but it will not harm to give 1-2 days to fix the bug. as this was in develop for x months (three months ) it can be concluded that its severity is very low, and this bug was not even detected by the e2e from this be can conclude that the place is not even tested properly by the e2e developers or the place is less visited by the developers, so the place of bug is not that imp and not harming other most developers, considering all this situation I must say the contributor must have 1-2 days to fix the bug not focusing on the reverting (some time solution is easy but it requires some debugging).


if the pr is not breaking the whole develop and e2e tests are passing on develop then we should at least give him some hours to fix it so that he can work peacefully and think about the solution.

**Comment by U8NWXD** [comment]
In the example I gave, I made the bug severe for the sake of argument (in that case the bug was time-dependent so it only started being severe after some period of time, which is why we wouldn't have noticed it earlier).

> as this was in develop for x months (three months ) it can be concluded that its severity is very low

If the bug is minor or moderate and only affecting developers, then exception 2 would apply and we could decide to give the contributor 2-7 days to fix it. I think this exception describes the scenario you're worried about, correct?
